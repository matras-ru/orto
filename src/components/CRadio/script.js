import { selfInstall } from '@/utils';
import commonAttributes from '@/mixins/commonAttributes.js';

import ThemeClass from '@/themes/default/CRadio';

const {
    disabledClass,
    baseIconClass,
    defaultIconClass,
    checkedIconClass,
    disabledIconClass,
    labelClass,
    inputClass,
    wrapperClass
} = ThemeClass;

export default {
    name: 'Radio',

    install(Vue, theme) {
        selfInstall(Vue, theme, this);
    },

    props: {
        ...commonAttributes.props,
        label: {
            type: String,
            default: null
        },
        modelValue: {
            type: [String, Number, Boolean],
            default: null
        },
        value: {
            type: [String, Number, Boolean],
            default: null
        }
    },

    model: {
        prop: 'modelValue',
        event: 'change'
    },

    data() {
        return {
            isChecked: false
        };
    },

    computed: {
        currentClass() {
            let classes = [labelClass];

            if (this.disabled) {
                classes.push(disabledClass);
            }

            return classes;
        },
        currentIconClass() {
            let classes = [baseIconClass];

            if (this.disabled) {
                classes.push(disabledIconClass);
            } else if (this.isChecked || this.shouldBeChecked) {
                classes.push(checkedIconClass);
            } else {
                classes.push(defaultIconClass);
            }

            return classes;
        },
        shouldBeChecked() {
            return this.modelValue === this.value;
        }
    },

    methods: {
        onChange(e) {
            this.$emit('change', e.target.checked ? this.value : null);
        }
    },

    render(h) {
        const attrs = {
            id: this.id,
            autofocus: this.autofocus,
            name: this.name,
            type: 'radio',
            disabled: this.disabled
        };

        const domProps = {
            checked: this.shouldBeChecked,
            value: this.value
        };

        const on = {
            change: this.onChange
        };

        const componentData = {
            class: inputClass,
            attrs,
            domProps,
            on
        };

        return h('label', { class: this.currentClass, attrs: { for: this.id } }, [
            h('input', componentData),
            h('div', { class: wrapperClass }, [
                h('span', { class: this.currentIconClass }),
                h('span', this.label)
            ])
        ]);
    }
};
