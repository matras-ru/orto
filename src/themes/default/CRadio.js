const Radio = {
    disabledClass: 'text-tertiary-200',
    baseIconClass: 'w-0-8 h-0-8 border-2 mr-0-5 mt-0-2 rounded-full',
    defaultIconClass: 'border-black-200 bg-white',
    checkedIconClass: 'border-black-200 bg-secondary-200 shadow-inner',
    disabledIconClass: 'border-tertiary-200 bg-white'
};

export default Radio;
