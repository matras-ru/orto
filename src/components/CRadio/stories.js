import { storiesOf } from '@storybook/vue';
import { action } from '@storybook/addon-actions';
import { withKnobs, text, boolean } from '@storybook/addon-knobs';

import CRadio from './';

const stories = storiesOf('CRadio', module);
stories.addDecorator(withKnobs);

stories
    .add('Radio knobs', () => ({
        components: { CRadio },
        props: {
            id: {
                default: text('id', 'radio1')
            },
            autofocus: {
                default: boolean('autofocus', false)
            },
            label: {
                default: text('label', 'Radio1')
            },
            name: {
                default: text('name', 'radio1')
            },
            disabled: {
                default: boolean('disabled', false)
            }
        },
        template:
            '<CRadio @onChange="action" :id="id" :autofocus="autofocus" :name="name" :label="label" :disabled="disabled" />',
        methods: { action: action('change') }
    }))
    .add('Default radio', () => ({
        components: { CRadio },
        template: '<CRadio @onChanget="action" id="radio2" label="Radio2" name="radio2" />',
        methods: { action: action('change') }
    }))
    .add('Disabled radio', () => ({
        components: { CRadio },
        template: '<CRadio @onChange="action" id="radio3" label="Radio3" name="radio3" disabled />',
        methods: { action: action('change') }
    }));
