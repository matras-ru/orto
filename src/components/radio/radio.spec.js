/* eslint-env jest */
import { mount, createLocalVue } from '@vue/test-utils';
import CRadio from './radio';

describe('Radio', () => {
    it('is rendered', () => {
        const wrapper = mount(CRadio, {
            context: {
                props: {
                    id: 'radio1',
                    name: 'radio1',
                    value: 'radio1',
                    label: 'default'
                }
            }
        });

        const input = wrapper.find('input');
        const label = wrapper.find('label');
        const icon = wrapper.find('span');

        expect(input.attributes('id')).toBeDefined();
        expect(input.attributes('id')).toBe('radio1');
        expect(input.attributes('type')).toBeDefined();
        expect(input.attributes('type')).toBe('radio');
        expect(input.attributes('name')).toBeDefined();
        expect(input.attributes('name')).toBe('radio1');

        expect(input.attributes('name')).toBe('radio1');

        expect(label.attributes('for')).toBe('radio1');
        expect(wrapper.text()).toBe('default');

        expect(wrapper.classes().sort()).toEqual('flex flex-wrap mb-0-7'.split(' ').sort());
        expect(input.classes().sort()).toEqual('absolute opacity-0 invisible'.split(' ').sort());
        expect(label.classes().sort()).toEqual('inline-flex cursor-pointer'.split(' ').sort());
        expect(icon.classes().sort()).toEqual(
            'w-0-8 h-0-8 border-2 mr-0-5 mt-0-2 rounded-full border-black-200 bg-white'
                .split(' ')
                .sort()
        );
    });

    it('is disabled', () => {
        const wrapper = mount(CRadio, {
            context: {
                props: {
                    id: 'radio1',
                    name: 'radio1',
                    value: 'radio1',
                    disabled: true
                }
            }
        });

        const input = wrapper.find('input');
        const label = wrapper.find('label');
        const icon = wrapper.find('span');

        expect(input.attributes('disabled')).toBeDefined();
        expect(input.attributes('disabled')).toBe('disabled');

        expect(label.classes().sort()).toEqual('inline-flex cursor-not-allowed'.split(' ').sort());
        expect(icon.classes().sort()).toEqual(
            'w-0-8 h-0-8 border-2 mr-0-5 mt-0-2 rounded-full border-tertiary-200 bg-white'
                .split(' ')
                .sort()
        );
    });

    it('is error', () => {
        const wrapper = mount(CRadio, {
            context: {
                props: {
                    id: 'radio1',
                    name: 'radio1',
                    value: 'radio1',
                    error: true
                }
            }
        });

        const icon = wrapper.find('span');

        expect(icon.classes().sort()).toEqual(
            'w-0-8 h-0-8 border-2 mr-0-5 mt-0-2 rounded-full border-danger bg-white'
                .split(' ')
                .sort()
        );
    });

    it('is checked', () => {
        const wrapper = mount(CRadio, {
            context: {
                props: {
                    id: 'radio1',
                    name: 'radio1',
                    value: 'radio1',
                    modelValue: 'radio1'
                }
            }
        });

        const label = wrapper.find('label');
        const icon = wrapper.find('span');
        expect(label.classes().sort()).toEqual('inline-flex cursor-pointer'.split(' ').sort());
        expect(icon.classes().sort()).toEqual(
            'w-0-8 h-0-8 border-2 mr-0-5 mt-0-2 rounded-full border-black-200 bg-secondary-200 shadow-inner'
                .split(' ')
                .sort()
        );
    });

    it('v-model', () => {
        const localVue = new createLocalVue();

        const App = localVue.extend({
            data() {
                return {
                    model: null
                };
            },

            render(h) {
                return h(CRadio, {
                    props: {
                        id: 'radio1',
                        name: 'radio1',
                        value: 'radio1',
                        modelValue: this.model
                    },
                    on: {
                        change: val => {
                            this.model = val;
                        }
                    }
                });
            }
        });

        const wrapper = mount(App, {
            localVue: localVue
        });

        const input = wrapper.find('input');

        expect(wrapper.vm.model).toEqual(null);

        input.element.selected = true;
        input.trigger('click');

        expect(wrapper.vm.model).toEqual('radio1');

        wrapper.destroy();
    });
});
