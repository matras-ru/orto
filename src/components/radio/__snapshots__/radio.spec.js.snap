// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Radio is disabled 1`] = `<div class="flex flex-wrap mb-0-4"><label for="radio1" class="relative pl-1-4 cursor-not-allowed opacity-50"><input id="radio1" name="radio1" type="radio" disabled="disabled" class="absolute top-0-2 left-0 w-0-8 h-0-8 form-radio" value="radio1"></label></div>`;

exports[`Radio is error 1`] = `<div class="flex flex-wrap mb-0-4"><label for="radio1" class="relative pl-1-4 cursor-pointer"><input id="radio1" name="radio1" type="radio" class="absolute top-0-2 left-0 w-0-8 h-0-8 form-radio form-radio-is-error" value="radio1"></label></div>`;

exports[`Radio is rendered 1`] = `<div class="flex flex-wrap mb-0-4"><label for="radio1" class="relative pl-1-4 cursor-pointer"><input id="radio1" name="radio1" type="radio" class="absolute top-0-2 left-0 w-0-8 h-0-8 form-radio" value="radio1">default</label></div>`;
